### DISCUSSION TITLE
Future Development Crew / Mirai

### DISCUSSION THREAD

**Discussion by shikokuchuo** [r-lib/mirai#60] [Ideas]
Not to propose a new direction in development, but this does get me thinking big-picture about the `mirai`/`crew` integration. `mirai` has a dispatcher as a separate process which frees the user session. But in `crew`, because of auto-scaling and transient workers, the main R process needs to be an event loop that continuously calls `push()`, `pop()`, or `wait()` ([similar to Gabor's task queue idea for `callr`](https://www.tidyverse.org/blog/2019/09/callr-task-q/)). The `crew` event loop is even more important now that there is throttling. All this makes me wonder if `crew` itself could manually run an iteration of dispatch each time e.g. `pop()` is called, rather than overburdening the `mirai` dispatcher. It might not be feasible due to the scope and intent of `mirai`, but I keep wondering about this.

_Originally posted by @wlandau in https://github.com/wlandau/crew/issues/76#issuecomment-1551827347_

**Comment by shikokuchuo** [comment]
On this I had an idea. I am prepared to provide access to dispatcher through a *language* interface.

By that I mean an argument at `dispatcher()` that takes a language object, or list of language objects, that can be evaluated in the context of the function itself at appropriate points in dispatcher's event loop.

As you know, the dispatcher event loop is hyper-efficient as it really is event-driven rather than polling. This means it is simply *waiting* most of the time. This is only possible due to its use of synchronisation primitives that `nanonext` makes available from NNG.

In this way, the additional scaling and launching logic can be retained within `crew`, and `mirai` can stick with its minimalist design purpose.

I would be happy to facilitate this if you find this appealing. I use this type of interface all the time in my machine learning work, so it's pretty much second nature to me. The ability for meta-programming / DSL is one of the core strengths of the R language and it would be rather cool to make use of it here. 

@wlandau

**Reply by wlandau** [reply]
Thanks for mulling this over. I see it as very long-term, and I think it is good that we are pausing and discussing first in order to make sure an implementation would truly be worth your time.

In your idea, would the dispatcher be part of the main local thread instead of a separate process? If the interface you propose would e.g. allow me to inject code into `dispatcher()` locally to get the output of `daemons()`, then that would fit right into `crew`.

My next comment is probably redundant, but I just want to make sure we are on the same page. The dispatcher in `mirai` is magic: extremely fast, event-driven, and completely non-blocking. And yet because of its own specific needs, `crew` itself has a dispatcher of its own (for servers instead of tasks). What's more, the `crew` dispatcher functionality requires a local event loop which blocks the main R session. I think this aspect of `crew` means that it is not getting the full benefits of the `mirai` dispatcher and that there is some effort duplicated over both packages. So in my original post, I was wondering if `crew` could take on responsibilities in its own local event loop that the current `mirai` dispatcher assumes in its external R process.

**Reply by shikokuchuo** [reply]
Yes. Keeping things high-level. I am saying `crew` can inject its own code to run in the dispatcher process (the same separate process it is now). That way both packages share the same event loop. As your code would run within the function, you would have access to all the internal variables such as `activevec`, `serverfree` etc. such that you would not need to query anything through a function interface such as `daemons()`.

While I will likely develop the threaded approach as well at some point, this may also allow `crew` to access for example the task queue directly. However, paradoxically it may be more difficult to integrate code that runs asynchronously, as anything that runs on a thread will have to be coded entirely in C (you can't call the R interpreter from another thread).

**Reply by wlandau** [reply]
> Yes. Keeping things high-level. I am saying crew can inject its own code to run in the dispatcher process (the same separate process it is now). That way both packages share the same event loop. As your code would run within the function, you would have access to all the internal variables such as activevec, serverfree etc. such that you would not need to query anything through a function interface such as daemons().

With functionality like this, I think it may be possible to inject auto-scaling logic directly into the dispatcher. It could be a heavy refactor for both of us, and there may or may not be breaking changes in `crew`. But if it works, it would allow `crew` to operate in a non-blocking way.

> While I will likely develop the threaded approach as well at some point, this may also allow crew to access for example the task queue directly. However, paradoxically it may be more difficult to integrate code that runs asynchronously, as anything that runs on a thread will have to be coded entirely in C (you can't call the R interpreter from another thread).

The `crew` [launcher plugin](https://wlandau.github.io/crew/articles/launcher_plugins.html) interface is based in R code, and I don't think there would be a way to work around that at the C level to make `crew` non-blocking. But this would not be a bad outcome for `targets` or Shiny, which I think are the major use cases. And with the dispatcher as a lighter-weight thread instead of a full R process, the consequences of duplicated effort would be mitigated.

Overall, both seem like potentially useful directions, but it seems like it would be good to ultimately commit to just one of these two approaches. On my end, it may take more time, experience, and user feedback before I am ready to request one way or the other.

**Reply by wlandau** [reply]
> On this I had an idea. I am prepared to provide access to dispatcher through a language interface.
> 
> By that I mean an argument at dispatcher() that takes a language object, or list of language objects, that can be evaluated in the context of the function itself at appropriate points in dispatcher's event loop.
> 
> As you know, the dispatcher event loop is hyper-efficient as it really is event-driven rather than polling. This means it is simply waiting most of the time. This is only possible due to its use of synchronisation primitives that nanonext makes available from NNG.
> 
> In this way, the additional scaling and launching logic can be retained within crew, and mirai can stick with its minimalist design purpose.

I reflected on this, and I think I am starting to wrap my head around it. At first I thought maybe I would have to split apart the `crew` controller and launcher somehow, which is tough because they need to talk to each other on the same process. But maybe I could have the controller create an altered version of itself (a “regulator” for just the workers rather than a full controller for both workers and tasks) to run on the `mirai` dispatcher, then ship it together with the launcher to the dispatcher process. These are moderate-sized objects, definitely not minimalist, but they would only need to be shipped once at the beginning.

This would be really rewarding if successful, and I would probably be able to make it back-compatible. I would be keen to try it if you are still willing to expose a language interface at some point. However, this direction would require a heavy refactor of `crew`, and I feel like I am at or near the end of the latest `crew` development cycle. So no  rush by any means, I may not be ready to even start experimenting for several weeks.

**Reply by wlandau** [reply]
On the other hand, maybe `crew`'s auto-scaling would get in the way of the efficiency of the dispatcher. With throttling on `crew`'s end, hopefully not, but at the moment I am not sure.

**Reply by shikokuchuo** [reply]
Currently I am also not sure which way things will land. The elegance of the language interface makes it appealing, but the threaded approach may turn out to have broader use cases. I think it's wise to dwell on this for a couple of months before re-visiting.

**Reply by wlandau** [reply]
Sounds like a solid plan.

In the meantime, I might implement the prep work in https://github.com/wlandau/crew/issues/84 because it will also improve the OOP design. I am a bit busier with other projects right now though, and it may take me a few weeks to set aside enough time to start.

If you were to implement the language interface for the dispatcher, would I be able to get the same integer matrix as the current `daemons()$daemons`, as well as the number of currently unresolved `mirai` tasks? And will I be able to send the dispatcher an `R6` object that automatically accepts those inputs and runs periodically in the event loop? These things should make the transition smooth.

**Reply by shikokuchuo** [reply]
The proposed timing sounds perfect.

You would have the daemons outputs, but at the moment I don't think you would have the unresolved tasks - as they could be buffered at the C level (with NNG). I think this is one of those design points, where `mirai` takes a decentralised approach and does not keep track of state, whereas it is trivial for `crew`.

Also when you say periodically, do you mean polling? I would be keen not to muddy the waters as dispatcher currently offers a clear demonstration of the benefits of the synchronisation primitives.

I am starting to think a jointly-developed package implementing this more advanced dispatcher may be a more flexible solution.

**Reply by wlandau** [reply]
> You would have the daemons outputs, but at the moment I don't think you would have the unresolved tasks - as they could be buffered at the C level (with NNG). I think this is one of those design points, where mirai takes a decentralised approach and does not keep track of state, whereas it is trivial for crew.

What kind of task-specific information is available? Presumably, I would see the number of tasks currently assigned to servers. Would I see the number of tasks that have not yet been dispatched? 

> Also when you say periodically, do you mean polling? I would be keen not to muddy the waters as dispatcher currently offers a clear demonstration of the benefits of the synchronisation primitives.

`crew` currently takes a polling-based approach, but maybe it wouldn't have to. What is the overall logical flow of the dispatcher event loop? I remember you said it spends most of its time sleeping, and I have noticed this makes `crew` super fast. Does it react to connections initiated by `mirai::mirai()` on the client? Is it involved in calls to `unresolved()` and `$data`? 

> I am starting to think a jointly-developed package implementing this more advanced dispatcher may be a more flexible solution.

Would it help if we set up some time to talk, maybe in late June?

**Reply by shikokuchuo** [reply]
> What kind of task-specific information is available? Presumably, I would see the number of tasks currently assigned to servers. Would I see the number of tasks that have not yet been dispatched?

Currently dispatcher only materialises from buffer messages it can handle (i.e. same as number of servers).  So there can be a lot of buffered incoming tasks that dispatcher does not know about yet.

> `crew` currently takes a polling-based approach, but maybe it wouldn't have to. What is the overall logical flow of the dispatcher event loop? I remember you said it spends most of its time sleeping, and I have noticed this makes `crew` super fast. Does it react to connections initiated by `mirai::mirai()` on the client? Is it involved in calls to `unresolved()` and `$data`?

Yes, essentially it kicks into action whenever (i) there is a new server connection, (ii) a task is received from the client (subject to the limitation above so this is not the same as every time a task is sent) or (iii) a completed task is received from a server. Also when `daemons()` status is called or `saisei()`. The rest of the time it just waits. `unresolved()` and `$data` would not trigger dispatcher.

> > I am starting to think a jointly-developed package implementing this more advanced dispatcher may be a more flexible solution.
> 
> Would it help if we set up some time to talk, maybe in late June?

Sounds like a good idea!

**Reply by wlandau** [reply]
> Yes, essentially it kicks into action whenever (i) there is a new server connection, (ii) a task is received from the client (subject to the limitation above so this is not the same as every time a task is sent) or (iii) a completed task is received from a server. Also when daemons() status is called or saisei(). The rest of the time it just waits. unresolved() and $data would not trigger dispatcher.

Thanks, this really helps me wrap my head around this.

> Sounds like a good idea!

Awesome! Please let me know when you are back from vacation and we can coordinate something.

**Reply by shikokuchuo** [reply]
Given the questions, I wonder if promises may actually be what you're looking for as they are designed precisely for side-effecty stuff you presumably want to do to scale workers. Example:

```r
library(promises)
library(mirai)
library(mirai.promises)
p1 <- as.promise(mirai({Sys.sleep(10); "complete"}))
p2 <- as.promise(mirai({Sys.sleep(10); "complete"}))
p1$then(cat)
p2$then(function() cat("\nhello"))
```
Or you may wish to call the side effect directly by modifying `as.promise.mirai()` in mirai.promises. You can see the logic of promises in that function in any case, you specify the eventual values for 'resolve' and 'reject', a check function that must not error or hang, and schedule the check function to poll using `later::later()`.

**Reply by wlandau** [reply]
There are definitely appealing aspects of `promises`/`later` for auto-scaling. I thought about `later` a few weeks ago, but it gets tricky on the client, e.g. if the user's R session blocks for an hour and delays auto-scaling that would otherwise run with only a 0.25-second throttling delay. If I could inject code into the dispatcher event loop, which sleeps almost all of the time, I could definitely try `later`:

1. When the dispatcher kicks into action, create an auto-scaling promise which will asynchronously resolve 0.25-seconds from now.
2. If the dispatcher kicks into action again before the promise resolves, do nothing.
3. When 0.25 seconds are up, resolve the promise next time the dispatcher event loop is available.
4. The next time the dispatcher wakes up and does something, create a new auto-scaling promise that asynchronously resolves later.
